services/command-router/src/main/java/org/eclipse/hono/commandrouter/impl/amqp/MappingAndDelegatingCommandHandler.java

```diff
@@ -112,124 +103,77 @@ public void mapAndDelegateIncomingCommandMessage(
             LOG.debug("command message address contains invalid tenant [expected: {}, found: {}]", tenantId, targetAddress.getTenantId());
             final Rejected rejected = new Rejected();
             rejected.setError(new ErrorCondition(AmqpError.UNAUTHORIZED_ACCESS, "unauthorized to send command to tenant"));
-            originalMessageDelivery.disposition(rejected, true);
+            messageDelivery.disposition(rejected, true);
             return;
-        } else if (deviceId == null) {
+        } else if (Strings.isNullOrEmpty(deviceId)) {
             LOG.debug("invalid command message address: {}", message.getAddress());
             final Rejected rejected = new Rejected();
             rejected.setError(new ErrorCondition(Constants.AMQP_BAD_REQUEST, "invalid command target address"));
-            originalMessageDelivery.disposition(rejected, true);
+            messageDelivery.disposition(rejected, true);
             return;
         }
 
-        final Command command = Command.from(message, tenantId, deviceId);
+        final ProtonBasedCommand command = ProtonBasedCommand.from(message);
         if (command.isValid()) {
-            LOG.trace("received valid command message: [{}]", command);
+            LOG.trace("received valid command message: {}", command);
         } else {
             LOG.debug("received invalid command message: {}", command);
         }
         final SpanContext spanContext = TracingHelper.extractSpanContext(connection.getTracer(), message);
         final Span currentSpan = CommandConsumer.createSpan("map and delegate command", tenantId, deviceId, null,
                 connection.getTracer(), spanContext);
-        CommandConsumer.logReceivedCommandToSpan(command, currentSpan);
-        final CommandContext commandContext = CommandContext.from(command, originalMessageDelivery, currentSpan);
+        command.logToSpan(currentSpan);
+        final ProtonBasedCommandContext commandContext = new ProtonBasedCommandContext(command, messageDelivery, currentSpan);
         if (command.isValid()) {
             mapAndDelegateIncomingCommand(tenantId, deviceId, commandContext);
         } else {
             // command message is invalid
-            commandContext.reject(getMalformedMessageError());
+            commandContext.reject("malformed command message");
         }
     }
 
-    private void mapAndDelegateIncomingCommand(final String tenantId, final String originalDeviceId,
-            final CommandContext originalCommandContext) {
-        final Command originalCommand = originalCommandContext.getCommand();
+    private void mapAndDelegateIncomingCommand(final String tenantId, final String deviceId,
+            final ProtonBasedCommandContext commandContext) {
+        final ProtonBasedCommand command = commandContext.getCommand();
```

Comment:
does this really need to be a `ProtonBasedCommand` instead of just a `Command`?

Reply:
`setGatewayId()` not being in the `Command` interface stood in the way.
But I've pulled that method into the interface now and added a legacy implementation.